Use 128 bit instead of Symbol for crate disambiguator

Xanewok opened the pull request:
As discussed on gitter, this changes `crate_disambiguator` from Strings to what they are represented as, a 128 bit number.

There's also one bit I think also needs to change, but wasn't 100% sure how: [create_root_def](https://github.com/rust-lang/rust/blob/f338dba29705e144bad8b2a675284538dd514896/src/librustc/hir/map/definitions.rs#L468-L482). Should I change `DefKey::root_parent_stable_hash` to accept `Fingerprint` as crate_disambiguator to quickly combine the hash of `crate_name` with the new 128 bit hash instead of a string for a disambiguator?

r? @michaelwoerister 

EDIT: Are those 3 tests `mir-opt` failing, because the hash is different, because we calculate it a little bit differently (storing directly instead of hashing the hex-string representation)? Should it be updated like in #45319?

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @michaelwoerister (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

Xanewok commented on src/librustc_incremental/persist/fs.rs:
```diff
@@ -614,21 +615,15 @@ fn string_to_timestamp(s: &str) -> Result<SystemTime, ()> {
     let crate_name = format!("{}-{}",
                              crate_name,
-                             base_n::encode(hasher.finish(), INT_ENCODE_BASE));
+                             base_n::encode(crate_disambiguator, INT_ENCODE_BASE));
```
This truncates the hash to 64 bits, should I implement a 128 but encoding in `base_n`?

arielb1 replied:
This is pre-existing?

Xanewok replied:
Now sure what you mean exactly, but [`to_smaller_hash()`](https://github.com/rust-lang/rust/blob/c0956ff265d0459e6f643f5454a8560e5dc2712c/src/librustc/ich/fingerprint.rs#L28) already exists and truncates to 64 bits.
I meant to ask if `base_n::encode` should be extended to work with `Fingerprint`/`u128`, in addition to `u64`. To be honest, I saw that it accepts `u64`, so thought that truncating 128->64 might be acceptable in this case

michaelwoerister replied:
The code you wrote should have roughly the same effect as before since `DefaultHasher` also produces 64 bits. We don't want these hashes to be too long because they end up in directory names and long paths/names sometimes cause trouble for various reasons (file system support, operating system API restrictions, etc).

Maybe you could add a comment here, similar to the one that is replaced:
```
// The full crate disambiguator is really long. 64 bits of it should be
// sufficient.
```

michaelwoerister commented on src/librustc/hir/map/collector.rs:
```diff
@@ -118,7 +119,7 @@ impl<'a, 'hir> NodeCollector<'a, 'hir> {
 
     pub(super) fn finalize_and_compute_crate_hash(self,
-                                                  crate_disambiguator: &str)
+                                                  crate_disambiguator: &Fingerprint)
```
Usually we pass around `Fingerprint` by value. It implements `Copy` so that should be easy.

michaelwoerister commented:
> There's also one bit I think also needs to change, but wasn't 100% sure how: create_root_def. Should I change DefKey::root_parent_stable_hash to accept Fingerprint as crate_disambiguator to quickly combine the hash of crate_name with the new 128 bit hash instead of a string for a disambiguator?

Yes, you can do that. It would be a bit more consistent this way.

Xanewok commented:
@michaelwoerister created a `CrateDisambiguator` newtype (awfully similar to [`DefPathHash`](https://github.com/rust-lang/rust/blob/f338dba29705e144bad8b2a675284538dd514896/src/librustc/hir/map/definitions.rs#L366)) in `session/mod.rs`. `config.rs` had a lot of CLI/config-related data and logic and it somehow felt better to put this small type directly inside `mod.rs`, especially since it's used in `Session` - hope that's okay.
Addressed comments and fixed tests :rocket:

michaelwoerister commented:
Excellent, thanks a lot @Xanewok! 

@bors r+

bors commented:
:pushpin: Commit 7fa64bc has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister 7fa64bcef3a23e2a19792ebe297ed17fab36c273 -->

bors commented:
:hourglass: Testing commit 7fa64bcef3a23e2a19792ebe297ed17fab36c273 with merge f764eaf453162cd19ef484ece07cc21e14dfb2c1...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5112), [status-travis](https://travis-ci.org/rust-lang/rust/builds/292595085?utm_source=github_status&utm_medium=notification)
Approved by: michaelwoerister
Pushing f764eaf453162cd19ef484ece07cc21e14dfb2c1 to master...

michaelwoerister commented:
🎉
